File: test/cli.test.ts
import { execFileSync } from "node:child_process";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import {
  dispatch,
  COMMANDS,
  IMPLEMENTATION_ID,
  MUTATING_COMMANDS,
  USAGE,
} from "../src/cli";

function captureOutput(fn: () => number) {
  let stdout = "";
  let stderr = "";
  const stdoutSpy = vi
    .spyOn(process.stdout, "write")
    .mockImplementation((chunk: unknown) => {
      stdout += String(chunk);
      return true;
    });
  const stderrSpy = vi
    .spyOn(process.stderr, "write")
    .mockImplementation((chunk: unknown) => {
      stderr += String(chunk);
      return true;
    });

  const exitCode = fn();

  stdoutSpy.mockRestore();
  stderrSpy.mockRestore();

  return { exitCode, stdout, stderr };
}

function git(cwd: string, args: string[]): void {
  execFileSync("git", args, { cwd, stdio: "pipe" });
}

describe("plan-sync CLI dispatch", () => {
  // Every dispatch call below resolves repoRoot/default-track from
  // process.cwd() (see src/repo-root.ts / src/sync-config.ts). Running these
  // tests from this project's own real repo root would read this project's
  // ACTUAL .omc/.sync-config.json (if one exists, e.g. from manual CLI
  // testing) and could execute REAL push/restore/status/uninstall against
  // real remote state — which is exactly what happened once, deleting a
  // real shadow-track setup. Every test here MUST run inside its own
  // fresh, isolated, no-config git repo.
  let tmpRoot: string;
  let anchorDir: string;
  let originalCwd: string;

  beforeEach(() => {
    tmpRoot = fs.mkdtempSync(path.join(os.tmpdir(), "omc-sync-cli-"));
    anchorDir = path.join(tmpRoot, "anchor");
    fs.mkdirSync(anchorDir, { recursive: true });
    git(anchorDir, ["init", "--quiet"]);

    originalCwd = process.cwd();
    process.chdir(anchorDir);
  });

  afterEach(() => {
    process.chdir(originalCwd);
    fs.rmSync(tmpRoot, { recursive: true, force: true });
  });

  it("prints usage listing all 7 subcommands and exits non-zero when given no args", () => {
    const { exitCode, stdout } = captureOutput(() => dispatch([]));

    expect(exitCode).not.toBe(0);
    expect(stdout).toContain(USAGE);
    for (const name of [
      "init",
      "allow",
      "unallow",
      "push",
      "pull",
      "status",
      "uninstall",
    ]) {
      expect(stdout).toContain(name);
    }
  });

  it("prints usage and exits non-zero for an unrecognized command", () => {
    const { exitCode, stdout } = captureOutput(() =>
      dispatch(["not-a-real-command"]),
    );

    expect(exitCode).not.toBe(0);
    expect(stdout).toContain(USAGE);
  });

  it("treats 'restore' as an unrecognized command now that pull covers both tracks", () => {
    const { exitCode, stdout } = captureOutput(() =>
      dispatch(["restore", "--track", "shadow"]),
    );

    expect(exitCode).not.toBe(0);
    expect(stdout).toContain(USAGE);
    expect(COMMANDS).not.toHaveProperty("restore");
  });

  it.each(["init", "allow", "unallow", "push", "pull", "status", "uninstall"])(
    "routes %s to its own command module (real error, not an unknown-command error)",
    (name) => {
      expect(COMMANDS).toHaveProperty(name);

      const { exitCode, stdout, stderr } = captureOutput(() =>
        dispatch([name]),
      );

      // With no --track, no persisted default track (fresh repo, no
      // .omc/.sync-config.json), and no prior init, every command is
      // expected to fail with a command-specific error — but it must NOT
      // be treated as an unrecognized command (no usage text), and the
      // error message should reference this specific command.
      expect(exitCode).not.toBe(0);
      expect(stdout).not.toContain(USAGE);
      expect(stderr).toContain(name);
    },
  );

  // The $PATH-collision identity marker (.omc/plans/go-port.md): with two
  // same-purpose binaries potentially on one $PATH, every MUTATING command
  // must announce which implementation ran it, on stderr, before anything
  // else it writes there.
  describe("stderr identity marker", () => {
    it.each([...MUTATING_COMMANDS])(
      "prints the identity marker as the first stderr line for the mutating command %s",
      (name) => {
        const { stderr } = captureOutput(() => dispatch([name]));

        expect(stderr).toContain(`plan-sync: ${IMPLEMENTATION_ID} (${name})\n`);
        // FIRST line, ahead of the command's own error output — so a user
        // whose $PATH has both binaries can attribute the failure.
        expect(stderr.split("\n")[0]).toBe(
          `plan-sync: ${IMPLEMENTATION_ID} (${name})`,
        );
      },
    );

    it("does not print the identity marker for the read-only status command", () => {
      const { stderr } = captureOutput(() => dispatch(["status"]));

      expect(MUTATING_COMMANDS.has("status")).toBe(false);
      expect(stderr).not.toContain(IMPLEMENTATION_ID);
    });

    it("does not print the identity marker for an unrecognized command", () => {
      const { stderr } = captureOutput(() => dispatch(["not-a-real-command"]));

      expect(stderr).not.toContain(IMPLEMENTATION_ID);
    });

    it("identifies this implementation as ts/<version>, distinguishably from the Go binary", () => {
      expect(IMPLEMENTATION_ID).toBe("ts/0.1.0");
      expect(IMPLEMENTATION_ID).not.toContain("go/");
    });
  });
});

File: go/internal/cli/cli_test.go
package cli

import (
	"errors"
	"os"
	"strings"
	"testing"
)

// captureOutput swaps os.Stdout/os.Stderr for temp files for the duration
// of fn and returns what was written to each. Both writeIdentityMarker and
// Dispatch reference os.Stdout/os.Stderr at CALL time, so replacing the
// package-level variables is enough — no plumbing of io.Writers into the
// production code is required (and adding such a seam purely for a test
// would be a worse trade).
func captureOutput(t *testing.T, fn func() int) (exitCode int, stdout, stderr string) {
	t.Helper()

	dir := t.TempDir()
	outFile, err := os.Create(dir + "/stdout")
	if err != nil {
		t.Fatalf("creating stdout capture file: %v", err)
	}
	errFile, err := os.Create(dir + "/stderr")
	if err != nil {
		t.Fatalf("creating stderr capture file: %v", err)
	}

	origOut, origErr := os.Stdout, os.Stderr
	os.Stdout, os.Stderr = outFile, errFile
	exitCode = fn()
	os.Stdout, os.Stderr = origOut, origErr

	if err := outFile.Close(); err != nil {
		t.Fatalf("closing stdout capture file: %v", err)
	}
	if err := errFile.Close(); err != nil {
		t.Fatalf("closing stderr capture file: %v", err)
	}

	outBytes, err := os.ReadFile(dir + "/stdout")
	if err != nil {
		t.Fatalf("reading stdout capture file: %v", err)
	}
	errBytes, err := os.ReadFile(dir + "/stderr")
	if err != nil {
		t.Fatalf("reading stderr capture file: %v", err)
	}
	return exitCode, string(outBytes), string(errBytes)
}

// registerStub installs a command that records that it ran and then fails,
// so the marker's position relative to the command's OWN stderr output is
// observable. Commands is a package-level map, so registration is undone on
// cleanup to keep tests independent.
func registerStub(t *testing.T, name string) *bool {
	t.Helper()
	ran := false
	Commands[name] = CommandFunc(func([]string) error {
		ran = true
		return errors.New("stub failure from " + name)
	})
	t.Cleanup(func() { delete(Commands, name) })
	return &ran
}

// TestDispatchWritesIdentityMarkerForMutatingCommands is the Go half of the
// $PATH-collision identity-marker requirement in .omc/plans/go-port.md: with
// two same-purpose binaries potentially on one $PATH, every mutating command
// must announce which implementation ran it, on stderr, ahead of anything
// else it writes there. Its TS counterpart lives in test/cli.test.ts.
func TestDispatchWritesIdentityMarkerForMutatingCommands(t *testing.T) {
	for _, name := range []string{"init", "allow", "unallow", "push", "pull"} {
		t.Run(name, func(t *testing.T) {
			if !mutatingCommands[name] {
				t.Fatalf("%q is not registered as a mutating command", name)
			}
			ran := registerStub(t, name)

			exitCode, _, stderr := captureOutput(t, func() int {
				return Dispatch([]string{name})
			})

			if !*ran {
				t.Fatalf("the %s command was never reached", name)
			}
			want := "plan-sync: " + ImplementationID + " (" + name + ")"
			if !strings.Contains(stderr, want+"\n") {
				t.Fatalf("stderr = %q, want it to contain %q", stderr, want)
			}
			// FIRST line, ahead of the command's own error output — so a
			// user whose $PATH has both binaries can attribute the failure.
			if first := strings.SplitN(stderr, "\n", 2)[0]; first != want {
				t.Fatalf("first stderr line = %q, want %q", first, want)
			}
			if !strings.Contains(stderr, "stub failure from "+name) {
				t.Fatalf("stderr = %q, want the command's own error too", stderr)
			}
			if exitCode == 0 {
				t.Fatal("exit code = 0, want non-zero for a failing command")
			}
		})
	}
}

// TestDispatchOmitsIdentityMarkerForStatus pins `status` as read-only: it
// mutates nothing, so it must not announce itself.
func TestDispatchOmitsIdentityMarkerForStatus(t *testing.T) {
	if mutatingCommands["status"] {
		t.Fatal("status must not be registered as a mutating command")
	}
	ran := registerStub(t, "status")

	_, _, stderr := captureOutput(t, func() int {
		return Dispatch([]string{"status"})
	})

	if !*ran {
		t.Fatal("the status command was never reached")
	}
	if strings.Contains(stderr, ImplementationID) {
		t.Fatalf("stderr = %q, want no identity marker", stderr)
	}
}

// TestDispatchOmitsIdentityMarkerForUnknownCommand covers the paths that
// never reach a command at all: nothing was mutated, so nothing announces
// itself.
func TestDispatchOmitsIdentityMarkerForUnknownCommand(t *testing.T) {
	for _, argv := range [][]string{nil, {"not-a-real-command"}, {"--help"}} {
		_, _, stderr := captureOutput(t, func() int { return Dispatch(argv) })
		if strings.Contains(stderr, ImplementationID) {
			t.Fatalf("Dispatch(%v): stderr = %q, want no identity marker", argv, stderr)
		}
	}
}

// TestImplementationIDIsDistinguishableFromTS guards the one property the
// marker exists for: `grep` must be able to tell the two implementations
// apart, and the format must not collide with the "plan-sync: <message>"
// error prefix used everywhere else.
func TestImplementationIDIsDistinguishableFromTS(t *testing.T) {
	if ImplementationID != "go/0.1.0" {
		t.Fatalf("ImplementationID = %q, want %q", ImplementationID, "go/0.1.0")
	}
	if strings.Contains(ImplementationID, "ts/") {
		t.Fatalf("ImplementationID = %q must not look like the TS binary's", ImplementationID)
	}
}
